gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java

```diff
@@ -0,0 +1,723 @@
+    final Map<BigQueryRpc.Option, ?> optionsMap = optionMap(options);
+    try {
+      Table answer = runWithRetries(new Callable<Table>() {
+        @Override
+        public Table call() {
+          return bigQueryRpc.getTable(tableId.dataset(), tableId.table(), optionsMap);
+        }
+      }, options().retryParams(), EXCEPTION_HANDLER);
+      return answer == null ? null : BaseTableInfo.fromPb(answer);
+    } catch (RetryHelper.RetryHelperException e) {
+      throw BigQueryException.translateAndThrow(e);
+    }
+  }
+
+  @Override
+  public Page<BaseTableInfo> listTables(String datasetId, TableListOption... options)
+      throws BigQueryException {
+    return listTables(datasetId, options(), optionMap(options));
+  }
+
+  @Override
+  public Page<BaseTableInfo> listTables(DatasetId datasetId, TableListOption... options)
+      throws BigQueryException {
+    return listTables(datasetId.dataset(), options(), optionMap(options));
+  }
+
+  private static Page<BaseTableInfo> listTables(final String datasetId, final BigQueryOptions
+      serviceOptions, final Map<BigQueryRpc.Option, ?> optionsMap) {
+    try {
+      BigQueryRpc.Tuple<String, Iterable<Table>> result =
+          runWithRetries(new Callable<BigQueryRpc.Tuple<String, Iterable<Table>>>() {
+            @Override
+            public BigQueryRpc.Tuple<String, Iterable<Table>> call() {
+              return serviceOptions.rpc().listTables(datasetId, optionsMap);
+            }
+          }, serviceOptions.retryParams(), EXCEPTION_HANDLER);
+      String cursor = result.x();
+      Iterable<BaseTableInfo> tables = Iterables.transform(result.y(),
+          BaseTableInfo.FROM_PB_FUNCTION);
+      return new PageImpl<>(new TablePageFetcher(datasetId, serviceOptions, cursor, optionsMap),
+          cursor, tables);
+    } catch (RetryHelper.RetryHelperException e) {
+      throw BigQueryException.translateAndThrow(e);
+    }
+  }
+
+  @Override
+  public InsertAllResponse insertAll(InsertAllRequest request) throws BigQueryException {
+    final TableId tableId = request.table();
+    final TableDataInsertAllRequest requestPb = new TableDataInsertAllRequest();
+    requestPb.setIgnoreUnknownValues(request.ignoreUnknownValues());
+    requestPb.setSkipInvalidRows(request.skipInvalidRows());
+    List<Rows> rowsPb = Lists.transform(request.rows(), new Function<RowToInsert, Rows>() {
+      @Override
+      public Rows apply(RowToInsert rowToInsert) {
+        return new Rows().setInsertId(rowToInsert.id()).setJson(rowToInsert.content());
+      }
+    });
+    requestPb.setRows(rowsPb);
+    try {
+      return InsertAllResponse.fromPb(runWithRetries(new Callable<TableDataInsertAllResponse>() {
+        @Override
+        public TableDataInsertAllResponse call() {
+          return bigQueryRpc.insertAll(tableId.dataset(), tableId.table(), requestPb);
+        }
+      }, options().retryParams(), EXCEPTION_HANDLER));
+    } catch (RetryHelper.RetryHelperException e) {
```

Comment:
One problem with retries here is that we can get duplicates if `insertId` is missing.

Some options:
1) document that when retries are configured make sure you provide id to avoid duplicates.

2) We should add a `UUID.randomUUID().toString()` when `insertId` is missing (before the `runWithRetries`) call.

3) Don't use retry handler here.